sys/usb/usbus/dfu/dfu.c

```diff
@@ -0,0 +1,292 @@
+/*
+ * Copyright (C) 2020 Mesotic SAS
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup usbus_dfu
+ * @{
+ * @file USBUS implementation for device firmware upgrade
+ *
+ *
+ * @author  Dylan Laduranty <dylan.laduranty@mesotic.com>
+ * @}
+ */
+
+#define USB_H_USER_IS_RIOT_INTERNAL
+
+#include "usb/dfu.h"
+#include "usb/descriptor.h"
+#include "usb/usbus.h"
+#include "usb/usbus/control.h"
+
+#ifdef MODULE_RIOTBOOT_USB_DFU
+#include "riotboot/flashwrite.h"
+#include "xtimer.h"
+#endif
+#include "periph/pm.h"
+
+#include "riotboot/slot.h"
+
+#include <string.h>
+
+#define ENABLE_DEBUG    (0)
+#include "debug.h"
+
+static void _event_handler(usbus_t *usbus, usbus_handler_t *handler,
+                          usbus_event_usb_t event);
+static int _control_handler(usbus_t *usbus, usbus_handler_t *handler,
+                            usbus_control_request_state_t state,
+                            usb_setup_t *setup);
+static void _transfer_handler(usbus_t *usbus, usbus_handler_t *handler,
+                              usbdev_ep_t *ep, usbus_event_transfer_t event);
+static void _init(usbus_t *usbus, usbus_handler_t *handler);
+
+#ifdef MODULE_RIOTBOOT_USB_DFU
+static void _reboot(void *arg);
+static bool skip_magic_word = true;
+static riotboot_flashwrite_t _writer;
+static char slot0_name[] = "RIOT-OS Slot 0";
+static char slot1_name[] = "RIOT-OS Slot 1";
+static bool ready = false;
+static xtimer_t scheduled_reboot = { .callback=_reboot };
+#define REBOOT_DELAY 2
+#else
+static char slot0_name[] = "RIOT-OS bootloader";
+#endif
+
+/* DFU relies on control endpoint only so it is safe to hardcode EP0 */
+#define dfu_ep usbus->ep_out[0].ep
+#define DEFAULT_XFER_SIZE 64
```

Comment:
I noticed that in a previous iteration, this depended on `FLASHPAGE_SIZE`, that's a bit dangerous because that's only defined for devices with a uniform flashpage layout. The stm32f4 devices will not have that define.

Reply:
In any case, default to 64 (endpoint size) works better for nrf52 so better drop the FLASHPAGE_SIZE argument here.